### ISSUE TITLE
Cache.putIfAbsent(K, V) contract is unusual (useless?) when used with a writer.

### ISSUE THREAD

**Issue by chrisdennis** [jsr107/jsr107spec#303]
The contract of putIfAbsent is:

> Atomically associates the specified key with the given value if it is not already associated with a value.
>
> This is equivalent to:
>```java
if (!cache.containsKey(key)) {}
  cache.put(key, value);
  return true;
} else {
  return false;
}
```
> except that the action is performed atomically.

This means that the 'ifAbsent' is a predicate not on the SoR state, but on the cache state.  This means that the put should only be performed if the mapping is absent from the cache, not if it is absent from the SoR.  To me this weird, I would expect that the desired behavior would be to act as a transparent caching proxy of the SoR in the presence of a cache writer.  As an example I would expect the following pseudocode to pass but per the letter of the specification it shouldn't and I believe in the RI it won't.
```java
sor.put(K, V1);
Cache cache = new Cache(sor);
assertThat(cache.putIfAbsent(K, V2), is(V1));
assertThat(cache.get(K), is(V1));
```

More details on this may be found under  ehcache/ehcache3#163

**Comment by chrisdennis** [comment]
@brianoliver - what kind of clarifications where you thinking of?  Clarifications of the method behavior in the documentation, changes to the contract, or something else entirely?

**Comment by cruftex** [comment]
> I actually think that use cases for a CacheWriter being used, also call for a CacheReader

No. There are always applications which are read-only, read/write and write-only in nature.

E.g. a write-only application could write logs or statistical data.

> What confuses me is that a .putIfAbsent isn't supposed to use the CacheLoader if one is configured, but should use the CacheWriter if that is there. So it works ignoring the CacheLoader, but not the CacheWriter.

The cache is doing what it is supposed to do: It is caching writes, or, otherwise put: Instead of optimizing on reads to a background system, it is optimizing write operations on a background system.

If the cache loader is invoked, there would always be an I/O operation to the background system, rendering the cache useless.

Furthermore, this method assumes that the application is the source of the data. There is no need to load it.

That sais, now, here is the big BUT:

Considered the fact, that this is a useful and nice method, the API is essentially incomplete:

- If there is a put() and a putIfAbsent() where is the getIfPresent() ?!
- Where is the putIfDifferent() ?!

The other, more general aspect of the discussion, is the fact that the cache configuration changes the semantics of the cache methods and this might be error prone and counter intuitive. I'll put this in another issue.

The cache API is not a consistent interface to a system of record. It is a cache API. We deal with two totally different requirements for an API: A data access API or an API to read and manipulate the cache state.

**Comment by alexsnaps** [comment]
On Tue, Jun 2, 2015 at 3:40 PM Jens Wilke <notifications@github.com> wrote:

> I actually think that use cases for a CacheWriter being used, also call
> for a CacheReader
>
> No. There are always applications which are read-only, read/write and
> write-only in nature.
>
> E.g. a write-only application could write logs or statistical data.
>
Wouldn't the later be a use-case for listeners instead? CacheLoader &
Writers are about integration with an underlying SoR afaict...


>  What confuses me is that a .putIfAbsent isn't supposed to use the
> CacheLoader if one is configured, but should use the CacheWriter if that is
> there. So it works ignoring the CacheLoader, but not the CacheWriter.
>
> The cache is doing what it is supposed to do: It is caching writes, or,
> otherwise put: Instead of optimizing on reads to a background system, it is
> optimizing write operations on a background system.
>
That is in the case of .put(), but what is the semantic of a .putIfAbsent()
then? So .get() would have returned, as it used the loader; as of now
though the .putIfAbsent() would succeed or not depending on what's in a
Cache; the problem is that you can't really control what's in the cache or
not. I don't think that the absence of something in a _Cache_ means
anything actually (other than it either hasn't been populated with that
mapping yet; or that the mapping wasn't considered "important" enough and
got evicted or expired).
As such I really can't see how someone programs around the .putIfAbsent()
contract, in the presence of CacheLoader/Writer, to be honest...



> If the cache loader is invoked, there would always be an I/O operation to
> the background system, rendering the cache useless.
>
No, only if the value is absent from the Cache. If it is there, there is no
need to retrieve it from the SoR. But if it is not present in the Cache,
but within the SoR, and the value present is being returned to me, now I
can program to that interface.


> Furthermore, this method assumes that the application is the source of the
> data. There is no need to load it.
>
> That sais, now, here is the big BUT:
>
> Considered the fact, that this is a useful and nice method, the API is
> essentially incomplete:
>
>    - If there is a put() and a putIfAbsent() where is the getIfPresent()
>    ?!
>
> I guess you .getIfPresent() is essentially a get ignoring the Loader if
the mapping isn't present?


>
>    - Where is the putIfDifferent() ?!
>
> I think in the alternate proposal, that would never happen. And the Cache
would just act as a fronting layer to the underlying SoR and always be
consistent. Which to me makes sense, as my understanding is that the
CacheLoader/Writer API is exactly about enabling the "cache through"
pattern, which pushes the concern of cache-to-SoR consistency concern into
the Cache.


>
>
> The other, more general aspect of the discussion, is the fact that the
> cache configuration changes the semantics of the cache methods and this
> might be error prone and counter intuitive. I'll put this in another issue.
>
Right, that's a falloff of that weird contract too though. In the
cache-through case (i.e. with consistent CacheLoader/Writer contract) you
get the same behavior as if the Cache would be containing your entire SoR
(and keeping the both in sync for you).


> The cache API is not a consistent interface to a system of record. It is a
> cache API. We deal with two totally different requirements for an API: A
> data access API or an API to read and manipulate the cache state.
>
In the Cache aside pattern totally, but again... CacheLoader/Writer are
about the Cache-through pattern... Don't you agree?

**Comment by cruftex** [comment]
> I guess you .getIfPresent() is essentially a get ignoring the Loader if the mapping isn't present?

Yes, in short: It is always ignoring the loader ;)

**Comment by alexsnaps** [comment]
But, well, then being present in the Cache has to have some sort of semantic the application can operate on. Which it doesn't, other than it didn't expire nor got evicted. But you can't infer anything from that. So what would be the value of these methods? What do you do any differently? ... I understand it would make it all more consistent... but merely with itself. I fail to see the added value to a user.

**Comment by cruftex** [comment]
> CacheLoader/Writer are about the Cache-through pattern... Don't you agree?

I agree, if you like to achieve perfection :)

However, that is difficult. E.g. in the Cache-Through pattern, should the method Cache.clear() not be available or should it consequently change its semantics to removeAll()?! But clearing the cache might be useful, isn't it? But when talking to an SoR, there should never be the need to clear the cache, right?

CacheLoader / CacheWriter is an incomplete interface for talking to an SoR. The CAS-operations do not work. And there is the thing called transactions.... But you can do still useful things with it!

**Comment by cruftex** [comment]
> But, well, then being present in the Cache has to have some sort of semantic the application can operate on. Which it doesn't, other than it didn't expire nor got evicted. But you can't infer anything from that. So what would be the value of these methods? What do you do any differently? ... I understand it would make it all more consistent... but merely with itself. I fail to see the added value to a user.

The cache is for improving performance, this leads to improved user experience. There are options for improving performance, by talking to the cache itself, so there is value.

Usecase 1: Conditional get support for a Java Servlet, serving cached data:

   protected long getLastModified(HttpServletRequest req) {
    String key = constructCacheKey(req);
    Content c = contentCache.getIfPresent(key);
    if (c != null) {
      return c.getMetaData().lastModifiedLong;
    }
    return -1;
  }

Usecase 2: Statistics, logging and debugging

      c = contentCache.getIfPresent(uri);
      if (c != null) {
        if (log.isDebugEnabled()) {
          log.debug("hit: " + uri);
        }
        res.setHeader("X-cache", "hit");
      } else {
      c = contentCache.get(uri);
        res.setHeader("X-cache", "miss");,
     }

Usecase 3: Go async, only if needed...

      c = content.getIfPresent(key);
      if (c != null) {
        doSomething(c);
      } else {
        load(key, callback to doSometing(c));
     }

**Comment by alexsnaps** [comment]
The first two use-cases aren't cache usage example are they? They are exposing a Cache's state through another API (i.e. some http proxy or something?) ... Confused. So it isn't surprising that these methods are somewhat "required". But I don't think that's the canonical use-case for a "Temporary Caching API for the Java platform" ...  

On the third one; how is that different than a `.get()` that would actually load the the value ?! i.e. 
```
c = cache.get(key); // always returns something, as it'll go to the SoR for you if absent
doSomething(c); 
```
Not sure what the `cache.load` method represent neither... Is the callback to be executed on a different thread or something?  

Also, the `null` now, in the CacheLoader/Writer case, has a different semantic (as always when you go from cache-aside to cache-through btw, so nothing surprising to a user here): it means nothing there... rather than nothing cached as of now...  Anyways, I'm slightly confused by the larger point you are trying to make here.

**Comment by cruftex** [comment]
> But I don't think that's the canonical use-case for a "Temporary Caching API for the Java platform" ...

That is the hopping point. The Caching API is for accessing the cache, it is not a data access API with the cache being transparent. If it would be a data access API, the method clear() is not allowed to exist.

You just need the data access API aspect only, if you just look on the aspect of software architecture and the intended semantics of your application over all. But the Caching API must address non functional requirements, too: Management, debugging, performance.

Besides your application there may be other tools and libraries needing to interact with your cache. E.g. an IoC container, or a management application to introspect the cache contents.

Another example:

      Cache.loadAll(keySet, false /* do not replace */, null /* don't notify */);

This may have an effect on the cache or not. Maybe now, maybe later something is loaded, if at all. There is no data flow to the application. So is it useless?

IMHO it is not useless, since the application can instruct the cache to load values that it needs at some point in time later on. Off topic: I would like to see this called Cache.prefetch(keySet).

What I am trying to show is:

I want to use a CacheLoader for various reasons, such as:
- I want to reconfigure the data source
- I want the cache take care that this is fresh data. E.g. the cache might also do refresh entries automatically in background, when expired. This means, the cache needs to know how to get the data

Using a loader does not imply that the cache is transparent by all means. There may still valid aspects to access the cache state itself.

**Comment by alexsnaps** [comment]
I meant point wrt to _this_ issue... what's your argument? Is it that the current contract is perfectly fine? And that this is what is to be expected of a "cache-through" cache?
I mean, the spec went final, many of the points I think you "argue" about aren't going to change at this stage, right...?

**Comment by cruftex** [comment]
> I'm slightly confused by the larger point you are trying to make here.

Maybe to "unconfuse": I can see exactly your point and where you are coming from. It would be a cleaner software architecture if there is an API that provides a transparent data access. This means there is nothing on the API which exposes the cache.

BTW: Consequently, a proper API design with this requirement would be layered or stackable. This leads to something like: Cache extends CacheLoader, CacheWriter.

From an architects point of view I may want an interface that is just for transparent data access. It makes pretty much sense to factor this part out. It could be a subset of the methods that is now on the Cache interface, so it is:

    class TransparentKeyValueAccess {
       V get(K);
       Map<K,V> getAll(Set<K>);
       void put(K, T);
       ...
    }

And Cache extends TransparentKeyValueAccess.

This could be added in maintenance without breaking the current interface. This way there is a cleaner API just for data access that cannot be misused. I think that is a good thing in terms of code quality in the applications.

**Comment by cruftex** [comment]
> I meant point wrt to this issue... what's your argument? Is it that the current contract is perfectly fine?

What I try is to sort and structure and address the various aspects that a "Caching API" inherently has, in a separate way. The API is a compromise to address the different aspects, there is no such thing then "perfect".

1. putIfAbsent() as it is now, has useful semantics.

2. In terms of naming it has correct semantics. "Absent" consequently addresses the concept of the cache state, since it is a caching API. If it would be otherwise we would have the identical discussion with containsKey(). It makes sense in the API that the wordings "absent", "contains", "(mapping) present" solely address the cache state. Everything different will be more confusing, not less.

> And that this is what is to be expected of a "cache-through" cache?

"cache-through" means loader and writer is present, correct? See point 2 above. The method is designed to interact upon the cache state. That is what "absent" tries to imply.

> I mean, the spec went final, many of the points I think you "argue" about aren't going to change at this stage, right...?

Now, you are confusing me!

You cannot have a perfect API, without working on it. Arguments are arguments no matter whether they can be addressed in a final standard now or in five years.

If I, personally, want to have a useful Caching API (in five years), I need to take care that useful stuff stays AND missing stuff gets in :)

Who is working on JCache 2.0 anyway, if not the people following and discussing here?

**Comment by cruftex** [comment]
So I tried to triage this ticket and distilled some aspects out of it:

- Missing operations that address the cache state: -> #319
- Confusing / improper behavior, with read/write through configurations: -> #318 
- Dito, confusing / improper behavior, with read/write through configurations: -> #317 

I think, these are tiny and feasible steps to address the underlying problems within the non perfect but useful world that we just have created.

**Comment by cruftex** [comment]
A left over: 

>> E.g. a write-only application could write logs or statistical data.

> Wouldn't the later be a use-case for listeners instead? CacheLoader &
Writers are about integration with an underlying SoR afaict...

Here we had a misunderstanding.

I meant that the applications purpose is to write statistical data. That written data has to be correct and, if it cannot be written, an error condition needs to be propagated, writes need to stall/block eventually for flow-control. So yes, in that use case we are interfacing to a SoR, but since the data is immutable in nature, there can be one application just writing and another just reading.

**Comment by cruftex** [comment]
Let me go back to the initial "issue". Brain was saying:

> I think the fundamental challenge is the putIfAbsent method itself as it essentially forces CacheLoaders and CacheWriters to be coordinated in-order to work in the way your use-case suggests.

That falls short.

1. The same way of thinking applies to all other cache methods that expose the cache state. These are many.
2. "coordination" is understatement: It has to be atomic to be correct and return a valid flag.

Simple and worst example: We cannot fix the method remove() for Alex' SoR purpose. To return the correct flag the CacheWriter needs to return a boolean, too.

Addressing Alex' use case is a valid request, but it is really big issue. Talking about the semantics of putIfAbsent is the tip of the iceberg. Here is the rest: Do we want to add CAS operations in the Loader and Writers and/or add transactions? Hi @brianoliver, you must have seen it coming!

**Comment by alexsnaps** [comment]
> Who is working on JCache 2.0 anyway, if not the people following and discussing here?

afaik, there is no scope defined yet and as such no JSR proposal in the making, nor an actual EG forming... I'm not entirely clear whether @gregrluck or @brianoliver would be interested in leading a 2.0 effort neither. (arguably it may still be a bit early for them to decide really too)... If not, maybe Cameron will! ;) 

That being said, you probably did well to create other issues for all these things you brought up. This issue is still targeted at `1.1` ... All these other issue _could_ be for a major... though, there too, some backwards compatibility is certainly desired.

**Comment by gregrluck** [comment]
The intent of CacheLoader/CacheWriter is not to create a transparent proxy. There would be a whole bunch of stuff you would change, not just this behaviour.

**Comment by chrisdennis** [comment]
Okay, so it's not intended to be a transparent proxy.  What, however, is it intended to be?  What use case is served by doing a putIfAbsent(K, V) on a cache with a writer given it's current contract?  Maybe this is something we could discuss at the BOF in SF?

**Comment by ljacomet** [comment]
+1 for having the discussion during the J1 BOF

**Comment by gregrluck** [comment]
Marked this as Not Fixing seeing as closing it does not work. 

We can discuss this at the BoF. My issue with is that similar considerations could apply to most of the atomic methods. 

The desired behaviour will be achieved if conditions are met:
a) the database or backing store is loaded into the cache so that it contains the whole population
b) the database or backing store is always kept in sync with the cache. 

In that case the atomic methods are perfectly usable. 

We do not want to make the cache spec slow. Needing to check a backing store would make it very slow and then the whole system loses usefulness.

**Comment by alexsnaps** [comment]
So sadly we didn't get to that level of details at the BOF, where @gregrluck wasn't there either anyways. 
I personally think (and so do our users, given the feedback we're getting) that the feature as it stands is ... at least _very_ confusing. On the other hand I do understand this isn't fixable in a maintenance release neither. 
I guess that for now, having this as "not fixing for 1.0" and hopefully address this in a subsequent release of the spec (i.e. a new JSR) is the best we can hope for. Not sure why the 2.0 milestone as a target was released though.

**Comment by cruftex** [comment]
While I agree that the brought up use cases are relevant to a cache and applications using it I suggest to close this issue. Reasons:

- We cannot change the existing behavior of `putIfAbsent` in a maintenance release of JSR107.
- The issue brings in new aspects of inter operations between loaders and writers and its possible consistency guarantees that are out of the scope of what is covered by JSR107 and which we can address here with the available resources

**Comment by gregrluck** [comment]
I think we agree this is not being fixed in a maintenance release for 1.1.
